boards/rp-pico/examples/pico_pwm_audio.rs

```diff
@@ -0,0 +1,244 @@
+    )
+    .ok()
+    .unwrap();
+
+    // The single-cycle I/O block controls our GPIO pins
+    let sio = hal::Sio::new(pac.SIO);
+
+    // Set the pins up according to their function on this particular board
+    let pins = rp_pico::Pins::new(
+        pac.IO_BANK0,
+        pac.PADS_BANK0,
+        sio.gpio_bank0,
+        &mut pac.RESETS,
+    );
+
+    // The delay object lets us wait for specified amounts of time (in
+    // milliseconds)
+    let mut delay = cortex_m::delay::Delay::new(core.SYST, clocks.system_clock.freq().integer());
+
+    // Init PWMs
+    let pwm_slices = hal::pwm::Slices::new(pac.PWM, &mut pac.RESETS);
+
+    // Setup the LED pin
+    let mut led_pin = pins.led.into_push_pull_output();
+
+    // Configure PWM0
+    let mut pwm = pwm_slices.pwm0;
+    pwm.default_config();
+
+    // 131,000,000 Hz divided by (top * div.int).
+    //
+    // fPWM = fSYS / ((TOP + 1) * (CSR_PH_CORRECT + 1) * (DIV_INT + (DIV_FRAC / 16)))
+    //
+    // 32kHz ~= 131,000,000 / ((4096 + 1) * 1 * 1)
+    pwm.set_top(4096);
+    pwm.set_div_int(1);
+
+    pwm.enable_interrupt();
+    pwm.enable();
+
+    // Output channel A on PWM0 to GPIO0
+    pwm.channel_a.output_to(pins.gpio0);
+
+    unsafe {
+        // Share the PWM with our interrupt routine.
+        PWM = Some(pwm);
+
+        // Unmask the PWM_IRQ_WRAP interrupt so we start receiving events.
+        pac::NVIC::unmask(pac::interrupt::PWM_IRQ_WRAP);
+    }
+
+    // N.B: Note that this would be much more efficiently implemented by using a DMA controller
+    // to continuously feed audio samples to the PWM straight from memory. The hardware
+    // is set up in a way where a rollover interrupt from the PWM channel can trigger a DMA
+    // request for the next byte (or u16) of memory.
+    // So while this is a good illustration for driving an audio signal from PWM, use DMA instead
+    // for a real project.
+
+    // Infinite loop, with LED on while audio is playing.
+    loop {
+        let _ = led_pin.set_high();
+
+        for i in AUDIO {
+            // Rescale from signed i8 numbers to 0..4096 (the TOP register we specified earlier)
+            //
+            // The PWM channel will increment an internal counter register, and if the counter is
+            // above or equal to this number, the PWM will output a logic high signal.
+            let i = ((*i as u16) << 4).wrapping_add(2048) & 0xFFF;
+
+            cortex_m::interrupt::free(|_| {
+                // SAFETY: Interrupt cannot currently use this while we're in a critical section.
+                let channel = &mut unsafe { PWM.as_mut() }.unwrap().channel_a;
+                channel.set_duty(i);
+            });
+
+            // Throttle until the PWM channel delivers us an interrupt saying it's done
+            // with this cycle (the internal counter wrapped). The interrupt handler will
+            // clear the interrupt and we'll send out the next sample.
+            cortex_m::asm::wfi();
```

Comment:
This only works if the PWM interrupt is the only one enabled.
Even for an example, I'd prefer a way which can be composed with other functionality.
Also, see [the documentation of WFI](https://developer.arm.com/documentation/dui0662/b/The-Cortex-M0--Instruction-Set/Miscellaneous-instructions/WFI): "WFI is intended for power saving only. When writing software assume that WFI might behave as a NOP operation."

Reply:
I agree, projects playing sounds are likely to have other hardware inputs such as ADC, buttons etc. so not ideal if the example stops working when other functionality is introduced.

Reply:
Ah good point. You guys think something like `wfe` and `sev` would be better instead?

Reply:
The approach used in @WeirdConstructor's code https://github.com/WeirdConstructor/RustRP2040Code/blob/master/pwm_dac_saw_sampling/src/example_1000hz_saw.rs which I think is the same as the pwm_blinky example seems much simpler. However there might be disadvantages I'm unaware of.